File: 18-4sum/18-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        
        numsLen = len(nums)
        
        if numsLen < 4: return []
        
        nums.sort()
        answer = []
        
        i = 0
        while i < numsLen - 3:
            
            if nums[i] > 0 and nums[i] > target:
                i += 1
                continue
                
            j = i + 1
            while j < numsLen - 2:
                
                twoSum = nums[i] + nums[j]
                
                if twoSum > 0 and twoSum > target: 
                    j += 1
                    continue
                
                left = j + 1
                right = numsLen - 1
                
                while left < right:
                    totalSum = twoSum + nums[left] + nums[right]
                    
                    if totalSum == target:
                        answer.append([nums[i], nums[j], nums[left], nums[right]])
                        
                        left += 1
                        while left < right and nums[left] == nums[left - 1]: 
                            left += 1
                            
                        right -= 1
                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1
                    elif totalSum > target:
                        right -= 1
                    else: 
                        left += 1
                        
                j += 1
                while j < numsLen - 2 and nums[j] == nums[j - 1]:
                    j += 1

            i += 1
            while i > 0 and i < numsLen - 3 and nums[i] == nums[i - 1]: 
                i += 1
                            
        return answer
                    
        
        
        

File: 18-4sum/18-4sum.cpp
class Solution {    // O(N^3)
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if (nums.size() < 4) return {};
        sort(nums.begin(),nums.end());
        vector<vector<int>> ret;
        
        for (int i = 0; i < nums.size() - 3; i++) {
            if (i && nums[i] == nums[i - 1]) continue;
            if (nums[i] > 0 && nums[i] > target) break;
            for (int j = i + 1; j < nums.size() - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                if (nums[i] + nums[j] > 0 && nums[i] + nums[j] > target) break;
                int left = j + 1, right = nums.size() - 1;
                while (left < right) {
                    long sum = (long)nums[i] + nums[j] + nums[left] + nums[right];
                    if (sum == target) {
                        ret.push_back({nums[i],nums[j],nums[left],nums[right]});
                        left++; right--;
                        while (left < right && nums[left] == nums[left - 1]) left++;
                        while (left < right && nums[right] == nums[right + 1]) right--; 
                    }
                    else if (sum > target) right--;
                    else left++;
                }
            }
        }
        return ret;
    }
};